lib/transcoder/audio.ex

```diff
@@ -90,32 +93,64 @@ defmodule Membrane.Transcoder.Audio do
     builder |> child(:aac_decoder, AAC.FDK.Decoder)
   end
 
-  defp maybe_plug_decoder(builder, %RawAudio{}) do
-    builder
+  defp maybe_plug_decoder(builder, %MPEGAudio{}) do
+    builder |> child(:mp3_decoder, Membrane.MP3.MAD.Decoder)
   end
 
-  defp maybe_plug_resampler(builder, %{sample_rate: sample_rate} = input_format, %Opus{})
-       when sample_rate != @opus_sample_rate do
-    builder
-    |> child(:resampler, %Membrane.FFmpeg.SWResample.Converter{
-      output_stream_format: %RawAudio{
-        sample_format: :s16le,
-        sample_rate: @opus_sample_rate,
-        channels: input_format.channels
-      }
-    })
+  defp maybe_plug_decoder(builder, %RemoteStream{content_format: MPEGAudio}) do
+    builder |> child(:mp3_decoder, Membrane.MP3.MAD.Decoder)
   end
 
-  defp maybe_plug_resampler(builder, %{sample_rate: sample_rate} = input_format, %AAC{})
-       when sample_rate not in @aac_sample_rates do
+  defp maybe_plug_decoder(builder, %RawAudio{}) do
     builder
-    |> child(:resampler, %Membrane.FFmpeg.SWResample.Converter{
-      output_stream_format: %RawAudio{
-        sample_format: :s16le,
-        sample_rate: 44_100,
-        channels: input_format.channels
-      }
-    })
+  end
+
+  defp maybe_plug_resampler(builder, input_format, %Opus{}) do
+    if Map.get(input_format, :sample_rate) == @opus_sample_rate do
+      builder
+    else
+      builder
+      |> child(:resampler, %Membrane.FFmpeg.SWResample.Converter{
+        output_stream_format: %RawAudio{
+          sample_format: :s16le,
+          sample_rate: @opus_sample_rate,
+          channels: Map.get(input_format, :channels, 1)
+        }
+      })
+    end
+  end
+
+  defp maybe_plug_resampler(builder, input_format, %AAC{}) do
+    if Map.get(input_format, :sample_rate) in @aac_sample_rates do
+      builder
+    else
+      builder
+      |> child(:resampler, %Membrane.FFmpeg.SWResample.Converter{
+        output_stream_format: %RawAudio{
+          sample_format: :s16le,
+          sample_rate: 44_100,
+          channels: Map.get(input_format, :channels, 1)
+        }
+      })
+    end
+  end
+
+  defp maybe_plug_resampler(
+         builder,
+         input_format,
+         %MPEGAudio{}
+       ) do
+    if Map.get(input_format, :sample_rate) == @mpeg_raw_audio_format.sample_rate and
+         Map.get(input_format, :sample_format) == @mpeg_raw_audio_format.sample_format and
+         Map.get(input_format, :channels) ==
+           @mpeg_raw_audio_format.channels do
+      builder
+    else
+      builder
+      |> child(:resampler, %Membrane.FFmpeg.SWResample.Converter{
+        output_stream_format: @mpeg_raw_audio_format
+      })
+    end
```

Comment:
Why have you deleted the guards in `maybe_plug_resampler/3`? You can just add `defp maybe_plug_resampler(builder, input_format, %MPEGAudio{})` with a proper guard to the old code and get rid off these `if` statements, it will make the code much prettier and easier to read IMO

Reply:
as discussed - sometimes there are no `sample_rate`, `sample_format` etc. fields in the `input_stream`

Reply:
Hm, ok, but should we plug resampler if some of these fields don't occur in the stream format?

Reply:
I've refactored it and added some custom guards